Support AllenNLPPruningCallback for AllenNLPExecutor

himkt opened the pull request:
## Motivation
This PR adds support for `AllenNLPPruningCallback` in `AllenNLPExecutor`.

## Description of the changes

- Implement `get_url` for storages
- Set some Optuna related information as environment variables in `AllenNLPExecutor`
- Load them in `AllenNLPPruningCallback`

I want to get feedback about (a) designing an interface, (b) adding internal APIs (`get_url`), and (c) usage of them.

himkt commented on optuna/storages/_redis.py:
```diff
@@ -65,6 +65,7 @@ def __init__(self, url: str) -> None:
 
         _imports.check()
 
+        self._url = url
```
In `RDBStororage`, `url` is a public attribute. Therefore I'm also wondering I should make it public.
https://github.com/optuna/optuna/blob/20ad7140c1bf492c45cb056593eecf3af4fee77f/optuna/storages/_rdb/storage.py#L104

toshihikoyanase replied:
If we provide the `get_url` method, I think `_url` can be private so that we can encourage users to use the method instead of the property.

himkt commented on optuna/integration/allennlp.py:
```diff
@@ -188,3 +224,11 @@ def __call__(
         self._trial.report(float(value), epoch)
         if self._trial.should_prune():
```
[TODO] The current implementation re-creates a pruner (`MedianPruner`, which is a default pruning algorithm in Optuna) even if we specify a pruner in Optuna side.

himkt replied:
https://github.com/optuna/optuna/pull/1772#discussion_r480194645
I implemented the API to restore a pruner defined in the Optuna side.

himkt commented on optuna/integration/allennlp.py:
```diff
@@ -157,22 +174,41 @@ class AllenNLPPruningCallback(EpochCallback):
-    def __init__(self, trial: optuna.trial.Trial, monitor: str):
+    def __init__(
+            self,
+            trial: Optional[optuna.trial.Trial] = None,
```
I want to make arguments `optional`, as `AllenNLPPruningCallback` would be initialized with any kind of information if it is specified in AllenNLP jsonnet configuration.

You can see the example in my GitHub repo.
https://github.com/himkt/optuna-allennlp/compare/master...feat/allennlp-jsonnet-pruning

himkt commented on optuna/integration/_allennlp.py:
```diff
@@ -0,0 +1,14 @@
+from typing import Callable
+
+
+class EpochCallback:
```
This is introduced for building documentation.

This PR adds a decorator `@EpochCallback.register` to `AllenNLPPruningCallback`,
which is undefined in document CI as it doesn't install `allennlp`.

crcrpar commented on optuna/integration/allennlp.py:
```diff
@@ -9,17 +10,33 @@
 else:
-    EpochCallback = object  # NOQA
+
+    class EpochCallback:  # type: ignore
```
We should explain why `# type: ignore` as I don't think it's evident at a glance.

himkt replied:
You're right. I added some notes in 5ef26c8.

crcrpar commented on optuna/integration/allennlp.py:
```diff
@@ -9,17 +10,33 @@
 
 if _imports.is_successful():
     import _jsonnet
+    from allennlp.training import EpochCallback
```
We should explain why only `EpochCallback` is imported conditionally.

himkt replied:
ditto.

crcrpar commented on optuna/integration/allennlp.py:
```diff
@@ -9,17 +10,127 @@
+
+    pruner_params = _get_environment_variables_for_pruner()
+
+    pruner = None  # type: Any
+    if pruner_class == "HyperbandPruner":
+        pruner = optuna.pruners.HyperbandPruner
+    elif pruner_class == "MedianPruner":
+        pruner = optuna.pruners.MedianPruner
+    elif pruner_class == "PercentilePruner":
+        pruner = optuna.pruners.PercentilePruner
+    elif pruner_class == "ThresholdPruner":
+        pruner = optuna.pruners.ThresholdPruner
+    elif pruner_class == "SuccessiveHalvingPruner":
+        pruner = optuna.pruners.SuccessiveHalvingPruner
+    elif pruner_class == "NopPruner":
+        pruner = optuna.pruners.NopPruner
```
Doesn't this work because this logic looks verbose to me?

```suggestion
    pruner = getattr(optuna.pruners, pruner_class, None)
```

himkt replied:
Amazing suggestion! Thank you.

crcrpar commented on optuna/integration/allennlp.py:
```diff
@@ -9,17 +10,127 @@
+        kwargs["min_resource"] = pruner._min_resource
+        kwargs["max_resource"] = pruner._max_resource
+        kwargs["reduction_factor"] = pruner._reduction_factor
+    elif isinstance(pruner, optuna.pruners.MedianPruner) or isinstance(
+        pruner, optuna.pruners.PercentilePruner
+    ):
```
Does this work?
```suggestion
    elif isinstance(pruner, optuna.pruners.PercentilePruner):
```

Ref: https://github.com/optuna/optuna/blob/fe1e3e54f961897b871daa7fe23d92c4e1eaa408/optuna/pruners/_median.py#L4

crcrpar reviewed (commented):
Dropping some comments that I don't think must be accessed :slightly_smiling_face:

toshihikoyanase commented on optuna/storages/_base.py:
```diff
@@ -558,6 +558,15 @@ def get_n_trials(self, study_id: int, state: Optional[TrialState] = None) -> int
         """
         raise NotImplementedError
 
+    @abc.abstractclassmethod
+    def get_url(self) -> Optional[str]:
+        """Get a path to a storage.
+
+        Returns:
+            A physical path to a storage.
+        """
+        raise NotImplementedError
```
The design of this method seems reasonable because of the following reasons:
1. `storage` argument of `optuna.create_study` accepts `str`, and all storages can have the URL string.
1. Some storages may not have the URL string (i.e., in-memory storage for Dask integration discussed in https://github.com/optuna/optuna/issues/1766), but such storage can raise errors.
1. Preserving `None` for `InMemoryStorage` is consistent with the other part of Optuna.

`BaseStorage` may be able to have `url` property, but I prefer the current proposal because of the second reason. Raising errors in the property does not seem to be a good practice.

This change will affect the public interface of the storage, and I'd like to hear opinions of developers and other reviewers, so your comments are more than welcome. (cc: @keisuke-umezawa @hvy @ytsmiling)

toshihikoyanase commented on optuna/storages/_cached_storage.py:
```diff
@@ -165,6 +165,12 @@ def get_all_study_summaries(self) -> List[StudySummary]:
+    def get_url(self) -> Optional[str]:
+
+        if hasattr(self, "_backend"):
+            return self._backend.url
```
A nit, but how about using the `get_url` for consistency?
```suggestion
            return self._backend.get_url()
```

himkt replied:
Ah, thank you. You're right. 🙇

toshihikoyanase reviewed (commented):
Thank you for your PR! I've heard that many AllenNLP users usually use config rather than Python script and this feature will be a great help of them.

In terms of the logic introduced in `AllenNLPExecutor`,  this PR employs a bit hacky way to use Optuna storages and pruners in AllenNLP processes, but it is acceptable because it mostly affects the code inside the integration module.

It also introduces a new `get_url` method to `optuna.storage.BaseStorage`, but I think it is simple and straightforward and it should have such interface as `RDBStorage` has the `url` property.

himkt commented on optuna/integration/allennlp.py:
```diff
@@ -157,22 +311,47 @@ class AllenNLPPruningCallback(EpochCallback):
-    def __init__(self, trial: optuna.trial.Trial, monitor: str):
+    def __init__(
+        self,
+        trial: Optional[optuna.trial.Trial] = None,
```
Making `trial` and `monitor` keyword argument so that `AllenNLPPruningCallback` can be instantiated without them. This change is introduced for making it possible to create a callback with environment variables. (https://github.com/optuna/optuna/pull/1772/files#diff-194650369fb014d2ea8f428774c04276R336-R350)

hvy replied:
We should probably explain in the documentation that these arguments are types optional but are mandatory.

himkt replied:
I added some explanations in 603308d. I'd like your feedback. :bow:

hvy replied:
Sorry if I wasn't clear but I meant in the user facing documentation.

himkt replied:
Thank you, it makes sense. :pray:
78822da I added a note in the docstring based on comments in the codebase.

crcrpar commented on optuna/integration/allennlp.py:
```diff
@@ -157,22 +328,47 @@ class AllenNLPPruningCallback(EpochCallback):
         monitor:
             An evaluation metric for pruning, e.g. ``validation_loss`` or
             ``validation_accuracy``.
+        trial:
+            A :class:`~optuna.trial.Trial` corresponding to the current evaluation of the
+            objective function.
```
`trial` should be put top of the `Args` block because the order of arguments of dunder init is trial and monitor.

himkt replied:
You're right, thank you! v

toshihikoyanase commented on optuna/integration/allennlp.py:
```diff
@@ -111,6 +254,26 @@ def __init__(
         else:
             self._include_package = include_package
 
+        system_attrs = {
```
As an alternative approach, we can get the storage URL without changing the storage API. How about separating the PR to implement `Storage.get_url()`?

```
storage = trial.study._storage
if isinstance(storage, RDBStorage):
    url = storage.url
elif isinstance(storage, RedisStorage):
    url = storage._url
elif isinstance(storage, _CachedStorage):
    assert isinstance(storage._backend, RDBStorage)
    url = storage._backend.url
else:
    raise ValueError("Storage must be RDBStorage or RedisStorage.")
```

himkt replied:
> How about separating the PR to implement Storage.get_url()?

Thank you for the suggestion, I agree with you.
I revised my PR (262f868 , 9b200af)  to remove `get_url` from each storage.

I rename `_url` in `RedisStorage` with `url` for consistency with `RDBStorage`. After this PR, I'll work on making `url` private in `RDBStorage` and `RedisStorage` at the same time.

Also, let me note `url` would be an empty string when no `url` can be loaded. It is because `AllenNLPExecutor` itself works on InMemoryStorage.

hvy replied:
Thanks a lot! Fewer files are now involved and it looks a bit cleaner. A small comment on the latest changes though, I think there's little benefit in aligning `RedisStorage`'s URL attribute to `RDBStorage` making it public. The `RDBStorage` has several of its attributes made public already and it's not just the `url`. We should probably limit the scope of the visibility matter to `RDBStorage`. It's always easier to make a private attribute public, than the other way around.

himkt replied:
Thank you for your suggestion. It makes sense. I made `url` private in `RedisStorage` at c537aa2.

hvy commented on optuna/integration/allennlp.py:
```diff
@@ -9,17 +10,142 @@
+            return wrapper
+
+
+PPID = os.getppid()
```
Is this constant intentionally defined public?

himkt replied:
I was not sure if I could put this on public space.
But now I made this private in 6af736b.

hvy commented on optuna/integration/allennlp.py:
```diff
@@ -163,16 +346,42 @@ class AllenNLPPruningCallback(EpochCallback):
+                self._monitor = monitor
+            else:
+                message = "Fail to load study.\n"
+                message += "AllenNLPPruningCallback works only with Optuna with RDB or Redis."
```
A nit. It might not be immediately obvious for the general users that "RDB" and "Redis" are alternatives to the `InMemoryStorage`.
```suggestion
                message += "AllenNLPPruningCallback works only with Optuna with RDB or Redis storages."
```
(Sorry I did not check the line length).

hvy commented on optuna/integration/allennlp.py:
```diff
@@ -9,17 +10,142 @@
+            return wrapper
+
+
+PPID = os.getppid()
```
A comment will surely help future maintainers, especially as you're using the parent id and not the id of the current process. Could you explain how this variable and the environment variables are used to communicate between processes?

himkt replied:
I created `_add_ppid` and added some notes there. 6af736b :pray:

hvy commented on optuna/integration/allennlp.py:
```diff
@@ -9,17 +10,142 @@
+        kwargs["min_resource"] = pruner._min_resource
+        kwargs["reduction_factor"] = pruner._reduction_factor
+        kwargs["min_early_stopping_rate"] = pruner._min_early_stopping_rate
+
```
I think we'll end up trying to re-instantiate a pruner with empty kwargs `{}` afterwards if an unrecognized pruner is given here. I don't think it's expected and we might want to raise a warning here in an `else`, if not an error.

himkt replied:
You're right. I added `else` in 0ddc47f.

himkt commented:
@hvy @toshihikoyanase 

Thank you for the reviews. I revised PR based on comments and suggestions.
PTAL.

toshihikoyanase commented on optuna/integration/allennlp.py:
```diff
@@ -163,16 +362,42 @@ class AllenNLPPruningCallback(EpochCallback):
+                self._trial = Trial(study, int(trial_id))
+                self._monitor = monitor
+            else:
+                message = "Fail to load study.\n"
+                message += "AllenNLPPruningCallback works only with Optuna and RDB or Redis storages."  # NOQA
```
Can we remove `# NOQA` as follows?
```suggestion
                message = (
                    "Fail to load study.\n"
                    "AllenNLPPruningCallback works only with Optuna and RDB or Redis storages."
                )
```

toshihikoyanase reviewed (commented):
Thank you for your update. I added some minor comments.
I'll check the coverage of test cases next.

himkt commented on optuna/integration/allennlp.py:
```diff
@@ -9,17 +10,158 @@
+    }
+
+
+def _get_environment_variables_for_pruner() -> Dict[str, Optional[str]]:
```
I'm sorry I didn't notice it. You're right.

hvy commented on optuna/integration/allennlp.py:
```diff
@@ -9,17 +10,156 @@
+        except ValueError:
+            if value == "True":
+                return True
+            elif value == "False":
```
A nit. Since we're returning if the preceding `if`.
```suggestion
            if value == "False":
```

hvy commented on optuna/integration/allennlp.py:
```diff
@@ -9,17 +10,156 @@
+
+def _get_environment_variables_for_trial() -> Dict[str, Optional[str]]:
+    return {
+        "study_name": os.getenv(_add_ppid("OPTUNA_ALLENNLP_STUDY_NAME")),
```
How about defining (most of) these constant strings at global scope? That'd prevent silly bugs introduced by typos. You can actually have them concatenated with the parent pid when constructing the constants.

himkt replied:
Thank you for the helpful suggestion. I defined parameters at the global scope in 65195ad.

toshihikoyanase commented on optuna/integration/allennlp.py:
```diff
@@ -9,17 +10,156 @@
+        kwargs["min_resource"] = pruner._min_resource
+        kwargs["reduction_factor"] = pruner._reduction_factor
+        kwargs["min_early_stopping_rate"] = pruner._min_early_stopping_rate
+    else:
```
`NopPruner` is not supported. Could you add it?
```suggestion
    elif isinstance(pruner, optuna.pruners.NopPruner):
        pass
    else:
```

toshihikoyanase commented on optuna/integration/allennlp.py:
```diff
@@ -9,17 +10,156 @@
+    pruner = trial.study.pruner
+    kwargs = {}  # type: Dict[str, Any]
+
+    if isinstance(pruner, optuna.pruners.HyperbandPruner):
+        kwargs["min_resource"] = pruner._min_resource
+        kwargs["max_resource"] = pruner._max_resource
+        kwargs["reduction_factor"] = pruner._reduction_factor
```
`HyperbandPruner` has some other states including `_pruners` and `_total_trial_allocation_budget` (https://github.com/optuna/optuna/blob/master/optuna/pruners/_hyperband.py#L143-L146). I guess it initializes the successive halving pruner every trial. Do you think if it affects optimization performance?

toshihikoyanase replied:
For example, the estimation of `max_resource` cannot be estimated correctly if the value of `max_resource` is `auto` and the resource (epoch) of each trial is different.

But, maybe this is an issue of `HyperbandPruner` and should be resolved by the pruner.

toshihikoyanase commented on tests/integration_tests/allennlp_tests/test_allennlp.py:
```diff
@@ -238,3 +238,58 @@ def objective(trial: optuna.Trial) -> float:
+    trial = optuna.trial.Trial(study, study._storage.create_new_trial(study._study_id))
+
+    with pytest.raises(ValueError):
+        optuna.integration.allennlp._fetch_pruner_config(trial)
```
Could you add a test case to check if `_fetch_pruner_config` does not raise errors when it receives supported pruners?

himkt replied:
In 07a9de9, I added a parametrized test. In this test, I found some bugs. Thank you so much for the suggestion.

toshihikoyanase reviewed (changes requested):
Thank you for your update. I added some minor comments on the pruners.

hvy commented on tests/integration_tests/allennlp_tests/test_allennlp.py:
```diff
@@ -238,3 +238,58 @@ def objective(trial: optuna.Trial) -> float:
+    trial = optuna.trial.Trial(study, study._storage.create_new_trial(study._study_id))
+
+    with pytest.raises(ValueError):
+        optuna.integration.allennlp._fetch_pruner_config(trial)
```
How about testing this using public API, i.e. the executor?

himkt replied:
I added `test_allennlp_pruning_callback_with_invalid_executor` in 07a9de9.

hvy commented on optuna/integration/allennlp.py:
```diff
@@ -163,16 +360,44 @@ class AllenNLPPruningCallback(EpochCallback):
+                    "Fail to load study.\n"
+                    "AllenNLPPruningCallback works only with Optuna and RDB or Redis storages."
+                )
+                raise Exception(message)
```
Can we test this logic?

hvy replied:
Can we provide more information and finer grained error handling by using a more concrete type of error?
```suggestion
                raise RuntimeException(message)
```

hvy commented on optuna/integration/allennlp.py:
```diff
@@ -163,16 +360,44 @@ class AllenNLPPruningCallback(EpochCallback):
+            study_name = _environment_variables["study_name"]
+            trial_id = _environment_variables["trial_id"]
+            monitor = _environment_variables["monitor"]
+            storage = _environment_variables["storage"] or None
```
A comment for `or None` would be helpful.

himkt replied:
I revised the PR to check if `storage` is `None` 8d227cd with a specific message.
What do you think about this change? Do you think it'd be better to add more comments?

hvy replied:
Looks good, thanks!

hvy commented on optuna/integration/allennlp.py:
```diff
@@ -163,16 +360,44 @@ class AllenNLPPruningCallback(EpochCallback):
+        if trial is not None and monitor is not None:
+            self._trial = trial
+            self._monitor = monitor
+        else:
```
I guess we can end up here when user don't pass `trial` or `monitor`. The current exception about storage types could be misleading. How about handling this case with a different exception such as `ValueError` ?

himkt replied:
I'm not sure if I understand your intention correctly, I added some comments in 603308d and 2987f7d.
I divided the error into the one for storage (https://github.com/optuna/optuna/pull/1772/files#diff-194650369fb014d2ea8f428774c04276R420-R425) and the other (https://github.com/optuna/optuna/pull/1772/files#diff-194650369fb014d2ea8f428774c04276R405-R413).

Please correct me know if I misunderstand your comment.

hvy reviewed (changes requested):
I left some additional comments. Could you take a look? A bit concerned that the logic is difficult to test but I guess it's difficult.

himkt commented on optuna/integration/allennlp.py:
```diff
@@ -9,17 +10,156 @@
+    class EpochCallback:  # type: ignore
+        """Stub for EpochCallback.
+
+        This class is introduced for documentation CI.
```
You're right, thank you. I removed the mention to `documentation` in c415b49.

hvy commented on optuna/integration/allennlp.py:
```diff
@@ -15,12 +18,166 @@
+            return wrapper
+
+
+PPID = os.getppid()
```
How about making this constant and the ones below private as we did before, by prefixing them with underscores?

himkt replied:
7b2c92f made `PPID` private. :pray:

toshihikoyanase commented on tests/integration_tests/allennlp_tests/test_allennlp.py:
```diff
@@ -238,3 +240,163 @@ def objective(trial: optuna.Trial) -> float:
         assert study.trials[0].value == 1.0
+
+
+@pytest.mark.parametrize(
+    "pruner_name,pruner_kwargs",
+    [
+        ("hyperband", {"min_resource": 3, "max_resource": 10, "reduction_factor": 5}),
+        ("median", {"n_startup_trials": 8, "n_warmup_steps": 1, "interval_steps": 3}),
+        ("noop", {}),
+        (
+            "percentile",
+            {"percentile": 50.0, "n_startup_trials": 10, "n_warmup_steps": 1, "interval_steps": 3},
+        ),
+        (
+            "successive_halving",
+            {"min_resource": 3, "reduction_factor": 5, "min_early_stopping_rate": 1},
+        ),
+        ("threshold", {"lower": 0.0, "upper": 1.0, "n_warmup_steps": 3, "interval_steps": 2}),
+    ],
+)
+def test_allennlp_pruning_callback_with_executor(
```
Can we simplify this test case by providing the pruner classes instead of the class name?
The following is just an example. Please fix the type hint of `pruner_class` if you employ this approach.

```python
@pytest.mark.parametrize(
    "pruner_class,pruner_kwargs",
    [
        (optuna.pruners.HyperbandPruner, {"min_resource": 3, "max_resource": 10, "reduction_factor": 5}),
        (optuna.pruners.MedianPruner, {"n_startup_trials": 8, "n_warmup_steps": 1, "interval_steps": 3}),
        (optuna.pruners.NopPruner, {}),
        (
            optuna.pruners.PercentilePruner,
            {"percentile": 50.0, "n_startup_trials": 10, "n_warmup_steps": 1, "interval_steps": 3},
        ),
        (
            optuna.pruners.SuccessiveHalvingPruner,
            {"min_resource": 3, "reduction_factor": 5, "min_early_stopping_rate": 1},
        ),
        (optuna.pruners.ThresholdPruner, {"lower": 0.0, "upper": 1.0, "n_warmup_steps": 3, "interval_steps": 2}),
    ],
)
def test_allennlp_pruning_callback_with_executor(
    pruner_class: Callable, pruner_kwargs: Dict[str, Union[int, float]]
) -> None:
    input_config_file = (
        "tests/integration_tests/allennlp_tests/example_with_executor_and_pruner.jsonnet"
    )

    def run_allennlp_executor(pruner: optuna.pruners.BasePruner) -> None:
        study = optuna.create_study(
            direction="maximize",
            pruner=pruner,
            storage=storage,
        )
        trial = optuna.trial.Trial(study, study._storage.create_new_trial(study._study_id))
        trial.suggest_float("DROPOUT", 0.0, 0.5)
        executor = optuna.integration.AllenNLPExecutor(trial, input_config_file, serialization_dir)
        executor.run()

    with tempfile.TemporaryDirectory() as tmp_dir:
        pruner_name = pruner_class.__name__
        os.mkdir(os.path.join(tmp_dir, pruner_name))
        storage = "sqlite:///" + os.path.join(tmp_dir, pruner_name, "result.db")
        serialization_dir = os.path.join(tmp_dir, pruner_name, "allennlp")

        pruner = pruner_class(**pruner_kwargs)
        run_allennlp_executor(pruner)
        ret_pruner = optuna.integration.allennlp._create_pruner()

        assert isinstance(ret_pruner, pruner_class)
        for key, value in pruner_kwargs.items():
            assert getattr(ret_pruner, "_{}".format(key)) != value

```

himkt replied:
It really makes sense! I applied your suggestion in 90f3eec, thank you.

toshihikoyanase commented on optuna/integration/allennlp.py:
```diff
@@ -15,12 +18,166 @@
+To avoid this hazard, we add ID of a parent process to each key of
+environment variables.
+"""
+PREFIX = "{}_OPTUNA_ALLENNLP".format(_PPID)
+MONITOR = "{}_MONITOR".format(PREFIX)
+PRUNER_CLASS = "{}_PRUNER_CLASS".format(PREFIX)
+PRUNER_KEYS = "{}_PRUNER_KEYS".format(PREFIX)
+STORAGE_NAME = "{}_STORAGE_NAME".format(PREFIX)
+STUDY_NAME = "{}_STUDY_NAME".format(PREFIX)
+TRIAL_ID = "{}_TRIAL_ID".format(PREFIX)
```
Can we make these variables private similar to `_PPID`?
```suggestion
_PREFIX = "{}_OPTUNA_ALLENNLP".format(_PPID)
_MONITOR = "{}_MONITOR".format(PREFIX)
_PRUNER_CLASS = "{}_PRUNER_CLASS".format(PREFIX)
_PRUNER_KEYS = "{}_PRUNER_KEYS".format(PREFIX)
_STORAGE_NAME = "{}_STORAGE_NAME".format(PREFIX)
_STUDY_NAME = "{}_STUDY_NAME".format(PREFIX)
_TRIAL_ID = "{}_TRIAL_ID".format(PREFIX)
```

himkt replied:
Sorry for not considering the suggestion https://github.com/optuna/optuna/pull/1772#discussion_r488378436 well. 🙇 
I made these constants private in 41e654f.

toshihikoyanase commented on optuna/integration/allennlp.py:
```diff
@@ -112,6 +286,37 @@ def __init__(
+            assert isinstance(storage._backend, optuna.storages.RDBStorage)
+            url = storage._backend.url
+        else:
+            url = ""
```
When I use the in-memory storage, I got the following error due to this line:
```console
[W 2020-09-16 15:24:13,416] Trial 0 failed because of the following error: ArgumentError("Could not parse rfc1738 URL from string ''")
```
I think it is not easy for users to find the cause since they do not provide the storage URL by themselves. Is it possible to tell users to use the `RDBStorage` or `RedisStorage` when they use in-memory storage?

himkt replied:
Thank you for the catch.
I had to check whether `storage` is an empty string or not.
In d3f8837, I update the PR for it.

Now the following message appears when users attempt to use in-memory storage.

```
RuntimeError: If you want to use AllenNLPExecutor and AllenNLPPruningCallback, you have to use RDB or Redis storage.
```

toshihikoyanase replied:
Thank you for your update. I confirmed the error message.
SGTM although the error handling is getting a bit complicated by https://github.com/optuna/optuna/commit/d3f8837bd69f76410433525e42133ae888e5e7f2.

hvy commented:
Will the usage of `optuna_pruner` be documented somewhere?

toshihikoyanase reviewed (approved):
LGTM.

@hvy Please merge after your review.

---
> Will the usage of `optuna_pruner` be documented somewhere?

I also think that we can add `optuna_pruner` to the configuration in `examples/allennlp/classifier.jsonnet`, but it's optional and not a strong opinion.

himkt commented on optuna/integration/allennlp.py:
```diff
@@ -15,12 +18,166 @@
+
+            """
+
+            def wrapper(subclass: Any, *args: Any, **kwargs: Any) -> None:
```
You're right, thank you!

himkt commented:
d8eb667 I added `optuna_pruner` to `examples/allennlp/classifier.jsonnet` and update `examples/allennlp/allennlp_jsonnet.py` to run optimization with Hyperband.
And also I added the reference to allennlp-guide in docstring in e133e17.

hvy reviewed (approved):
LGTM! Thanks for the long running effort. I would just say that the logic is a bit tricky and not necessarily robust to (internal) changes. I believe we can discuss alternatives such as pickling if needed.

hvy commented:
Updated the title a bit, hope you're fine with that 🙂

himkt commented:
I'd really appreciate patient reviews from all reviewers. Thank you so much.
I agree that the current implementation is hacky and fragile, I'll work on improving mechanism.
